### ISSUE TITLE
training code hang when running without NCCL_PROTO

### ISSUE THREAD

**Issue by jumanzi-dev** [NVIDIA/nccl#354]
Hi,

I'm trying to run tensorflow benchmark code(https://github.com/tensorflow/benchmarks.git -b cnn_tf_v1.15_compatible)
It works fine if I use 4gpu on a node, but make problem when I use more than 16gpu and it's getting hang without WARN message. 
When job hangs every gpu usage hit 100% so I guess its fall on deadlock. 
While i'm try to fix this problem, i found that if I set NCCL_PROTO with specific value(no matter it's LL, LL128, Simple or combination) it's working. 

So My question is,
- How's NCCL select protocol when there is no NCCL_PROTO value? could we use 3 protocol as docs?
- could you give any hint for why this hang issue keeps happen?

Thanks!

===
my environment is:
- cuda10, nccl 2.5.6-1
- horovod 0.18.0
- openmpi 4.0.0
- 4 gpu(V100) per node and using 4 node. (total gpu is 16)
- nvlink for intra-node and IB for inter-node.

What I've tested is:
- working when I set IS_P2P_DISABLE=1
- did p2pBandwidthLatencyTest but there is no issue
- install horovod not to use NCCL for ALL_REDUCE, then it works fine.
- it's working when I set NCCL_PROTO with combination of LL, LL128, Simple
-> it looks like this hang is due to protocol bugs during p2p communication.

**Comment by kwen2501** [comment]
Hi, just taking a wild guess -- could it be possible that you have different NCCL versions on different nodes?

**Comment by jumanzi-dev** [comment]
@kwen2501 
No, we run training job based on same docker image.

**Comment by sjeaugey** [comment]
From what you describe, it seems indeed that different ranks are switching protocols at different sizes. Forcing one protocol (as you did) would workaround the problem.

Now, the question is why. We try to synchronize everything that could affect the tuning between ranks before we compute when to use which protocol/algorithm, so as @kwen2501 mentioned, the one case we cannot solve is when you have different NCCL versions on different ranks. 

Now, there could also be something we didn't synchronize correctly. Among the things that could make different ranks switch at different sizes, the main one would be the hardware. Are all systems the same (same platform, number&type of GPUs, number&types of NICs, ...?).

Another thing would be environment variables which would be set on one node but not on the other (e.g. things in /etc/nccl.conf).

To help work on the issue, it would be good to reproduce it with the NCCL tests (https://github.com/nvidia/nccl-tests); that would help better characterize the problem and run more experiments.

**Comment by jumanzi-dev** [comment]
@kwen2501 @sjeaugey 
Thanks for your time and also quick reply.

I found this error happens only when not pass NCCL_PROTO environment value.

Based on the source code, I found the default value of NCCL_PROTO_LL128's value is set by specific HW condition.
Otherwise if I pass the NCCL_PROTO, then it will be set by whatever I'd like to.
https://github.com/NVIDIA/nccl/blob/299c554dccf923230321ad7495946543f3e9b457/src/graph/tuning.cc#L142-L157

So Likewise your advise, It may due to some HW variance between nodes.
For now I can run the code with NCCL_PROTO value and will fix HW spec.

I appreciate you guys. Thanks!
